### ISSUE TITLE
Next steps in code generation

### ISSUE THREAD

**Issue by kbrandwijk** [Urigo/graphql-cli#81]
The current boilerplates are a great starting point for GraphQL server development, but I want to address a couple of points that occur after the initial server is created using a boilerplate in this proposal.

## Splitting generated and user code
The current boilerplates contain a single source file, containing everything to get the server up and running. While this is a great getting started experience, it's also creates a problem during further development of the server. The code can roughly be divided into two parts: user code, which is mainly the implementation of the resolvers, and any other custom code, and the plumbing around it. I think it would be a good next step to split these up into two different files. This paves the way for the next steps in this proposal.

## Adding endpoints
Whenever a new endpoint is added to the server, it ends up as a 'project' in `.graphqlconfig` (we might want to rethink the term 'project'). Next, a new binding has to be added to the server context (and optionally pre-generated). This is currently a manual step. It would be great if when adding an endpoint to the server, it would be possible to update the server code to add another binding to the context. Having no user code in that file will help with this.
- [ ] TODO: Create an example based on current boilerplate

## Knowing the endpoint address
In the current setup, I don't see a way to easily retrieve Graphcool endpoint addresses. They are not part of `graphcool.yml`, and there is a relationship to environment variables, stages, etc. At the end of the story, the endpoint address needs to end up in the server file for the binding to work. 
- [ ] _This point needs further detailing_

## Refreshing schemas
When using either compile-time bundling (processing schema imports) or generated binding files, there should be an easy way to regenerate those when a schema is refreshed. There are a few different actions that can trigger this, and we need to think about a way to detect those. For Graphcool services, there's `graphcool deploy` that downloads the new backend schema. For 'custom' endpoints, there's `graphql get-schema` and when using GrAMPS, adding a new datasource, or updating a datasource package, would cause a new schema.

Bottom line is, the getting started experience is great, but I think we can provide additional tooling during the development phase to make the inception to production experience even greater.

cc @schickling @jlengstorf @marktani

**Comment by Urigo** [comment]
@kbrandwijk those are great ideas!

We've just released a new version and an alpha release channel for GraphQL CLI - `4.0.0-alpha.XXX`, which was a complete rewrite.
Some of those changes Included:
1. Using the latest [GraphQL Config](https://medium.com/the-guild/graphql-config-708dd0e5d15f) 
2. Using [GraphQL Code Generator](http://graphql-code-generator.com)
3. Using [GraphBack](https://github.com/aerogear/graphback) for generating the initial projects using templates and generating other things.

Checkout new instructions and the [migration guide](https://github.com/urigo/graphql-cli#migration-from-3xx-to-4xx) and we will be happy to hear your feedback during the alpha period!

Your ideas are spot on and I would love to meet and chat about those and how we should go about implementing them in the new architecture.

Would you be willing to meet over video chat for a short talk about your ideas?

You can reach out to me thought my email (public on my Github profile) or through our Discord channel: 
https://discord.gg/xud7bH

**Comment by Urigo** [comment]
I've posted the wrong Discord link, this is the correct one: https://discord.gg/xud7bH9

**Comment by wtrocki** [comment]
I went thru ideas and I think that graphql-cli mapps now all of them. We are always looking for feedback. Going to close this issue, but feel free to reopen
